File: SQP/compare_blocking.m
function [result, ai, bi] = compare_blocking(theta, p, A_inequation, working_inequation_A, b_inequation, working_inequation_b)
%COMPARE_BLOCKING 此处显示有关此函数的摘要
%   此处显示详细说明
%   初步的想法是，这里返回的是 (bi - ai' * theta)/(ai' * p) 的所有符合
%   ai' * p < 0 的结果，以向量的形式
%   现在来解决问题1.
%   初步想法是，把A_inequation传进来，然后其每行与working set中的{2,1}比较
%   有一样的就剔除出去，剩下的就是我所需要的不在working set中的constraints
%   现在来解决问题2.
%   就，简单地对比一下A_temp中剩下的行与 p 的乘积，挑出其中乘积小于零的

%   A_inequation 是所有不等式constraint的系数矩阵
%   working_inequation 是所有在working set中的不等式constraints的系数矩阵

% 将包含所有不等式constraints的系数的矩阵A换存到 A_temp 方便后面的操作
A_temp = A_inequation;
b_temp = b_inequation;
size_A = size(A_temp);

% 这里将那些包含着在working set中的不等式constraints的系数矩阵存到temp中
% 方便后面操作
size_working = size(working_inequation_A);
working_A = working_inequation_A;
working_b = working_inequation_b;


% 这个for之后，A_temp中将剩下不在当前working set中的不等式constraints的系数
% 所组成的矩阵
delete_position = zeros(size(working_b))';
for i = 1:size_working(1,1)
    for j = 1:size_A(1,1)
        
        % 只要扫描到working set中某一个跟当前的A中的行相同的行，就删除该行
        % 并同时也在 A 中将这行删除
        % 随即就可以不用管working set中的别的元素了
        if A_temp(j, :) == working_A(i, :)
            delete_position(1 ,i) = j ;
            break;
        end
    end
end
% 从全部不等式constraints中，删除这些working set中有的不等式constraints
if delete_position ~= zeros(size(working_b))'
    A_temp(delete_position, :) = [];
    b_temp(delete_position, :) = [];
end



% 经过这个for之后，A_temp, b_temp中的应该是，不在working set中且 ai' * p 小于零
% 的constraints，除此之外的都是零行
size_row = size(A_temp(1,:));
for i = 1:size(A_temp)
    result = A_temp(i,:) * p ;
    if result > 0
        A_temp(i, :) = zeros(size_row);
        b_temp(i, 1) = 0 ;
    end
end

% 将其中的零行删除，就剩下我们的目标不等式constraints了
A_temp(~any(A_temp, 2), :) = [];
b_temp(~any(b_temp, 2), :) = [];


result_temp = zeros(size(b_temp));

% 计算出其中的所有 ( bi - ai' * theta ) / (ai' * p)
for i = 1:size(b_temp)
    result_temp(i, 1) = (b_temp(i, 1) - A_temp(i, :) * theta)/(A_temp(i, :) * p);
end

%result = 0;
result = min(result_temp);
aim = find(result == min(result_temp));
ai = A_temp(aim, :);
bi = b_temp(aim, 1);

end

File: python/compare_blocking.py
'''
28.01.2021 by Muqian Chen
I would like to use this package to include two functions

1. calculation_alpha accepts the current parameter in the working set
to calculate alpha parameter used to multiply with step length for
reaching a gradient descent.


'''



import numpy as np

def calculation_alpha(A_inequation, A_working, b_inequation, b_working, step_length, theta):
    # firstly thinking about situation with only unequaled constraints
    [row, col] = np.shape(b_working)
    exist_index = np.zeros(0, dtype=float)
    for i in range(row):
        b_temp = b_working[i, :]
        A_temp = A_working[i, :]

        # find the index of the same parameters inside
        # unequaled constraints
        b_index = np.where(b_inequation == b_temp)
        A_index = np.where(A_inequation == A_temp)

        # determine the right index from the result
        for j in range(len(b_index)):
            if_exist = np.array([A_index == b_index[i]])
            if if_exist.any():
                exist_index = np.hstack((exist_index, b_index[i]))
        # save the right index in np.array exist_index

    # delete all in the working set existed constraints from all unequaled constraints
    # so that the rest parameters are all not in working set.
    A_inequation = np.delete(A_inequation, exist_index, axis=0)
    b_inequation = np.delete(b_inequation, exist_index, axis=0)

    # pick out the parameters that satisfy inequation a_i' * p < 0
    b_result = np.zeros((0, 1), dtype=float)
    A_result = np.zeros((0, 2), dtype=float)

    index_list = []
    # please be careful, I would like to make b_result as a column vector
    for i in range(len(b_inequation)):
        if np.dot(A_inequation[i, :], step_length) < 0:
            b_result = np.vstack((b_result, b_inequation[i, :]))
            A_result = np.vstack((A_result, A_inequation[i, :]))
            index_list.append(i)

    # if there are not any satisfied constraints parameter inside b_result
    # It could just be said that the alpha parameter is one.
    # else, just save the b_i - a_i'*theta / a_i' * p results in the list
    # calc_result and then get the minimal value
    calc_result = []
    length = len(b_result)
    print(np.shape(b_result))
    if length == 0:
        alpha = 1
        b_blocked = np.zeros((0, 1), dtype=float)
        A_blocked = np.zeros((0, 2), dtype=float)
        return alpha, b_blocked, A_blocked
    else:
        for i in range(len(b_result)):
            temp = (b_result[i, 0] - np.dot(A_result[i, :], theta)) / np.dot(A_result[i, :], step_length)
            calc_result.append(temp)
        min_calc = min(calc_result)
        min_index = calc_result.index(min_calc)
        if min_calc < 1:
            alpha = min_calc
            b_blocked = b_result[min_index, 0]
            A_blocked = A_result[min_index, :]
            # return the b_i and a_i' of blocking constraint
            return alpha, b_blocked, A_blocked
        else:
            alpha = 1
            b_blocked = np.zeros((0, 1), dtype=float)
            A_blocked = np.zeros((0, 2), dtype=float)
            return alpha,  b_blocked, A_blocked


